examples/apfl_example/client.py

```diff
@@ -1,41 +1,38 @@
 import argparse
 from pathlib import Path
-from typing import Sequence
+from typing import Tuple
 
 import flwr as fl
 import torch
+import torch.nn as nn
 from flwr.common.typing import Config
+from torch.nn.modules.loss import _Loss
+from torch.optim import Optimizer
+from torch.utils.data import DataLoader
 
 from examples.models.cnn_model import MnistNetWithBnAndFrozen
 from fl4health.clients.apfl_client import ApflClient
 from fl4health.model_bases.apfl_base import APFLModule
-from fl4health.parameter_exchange.layer_exchanger import FixedLayerExchanger
 from fl4health.utils.load_data import load_mnist_data
-from fl4health.utils.metrics import Accuracy, Metric
+from fl4health.utils.metrics import Accuracy
 from fl4health.utils.sampler import DirichletLabelBasedSampler
 
 
 class MnistApflClient(ApflClient):
-    def __init__(
-        self,
-        data_path: Path,
-        metrics: Sequence[Metric],
-        device: torch.device,
-    ) -> None:
-        super().__init__(data_path=data_path, metrics=metrics, device=device)
-
-    def setup_client(self, config: Config) -> None:
+    def get_data_loaders(self, config: Config) -> Tuple[DataLoader, DataLoader]:
         batch_size = self.narrow_config_type(config, "batch_size", int)
-        self.model: APFLModule = APFLModule(MnistNetWithBnAndFrozen()).to(self.device)
-        self.criterion = torch.nn.CrossEntropyLoss()
-        self.local_optimizer = torch.optim.AdamW(self.model.local_model.parameters(), lr=0.01)
-        self.global_optimizer = torch.optim.AdamW(self.model.global_model.parameters(), lr=0.01)
         sampler = DirichletLabelBasedSampler(list(range(10)), sample_percentage=0.75)
+        train_loader, val_loader, _ = load_mnist_data(self.data_path, batch_size, sampler)
+        return train_loader, val_loader
 
-        self.train_loader, self.val_loader, self.num_examples = load_mnist_data(self.data_path, batch_size, sampler)
-        self.parameter_exchanger = FixedLayerExchanger(self.model.layers_to_exchange())
+    def get_model(self, config: Config) -> nn.Module:
+        return APFLModule(MnistNetWithBnAndFrozen()).to(self.device)
 
-        super().setup_client(config)
+    def get_optimizer(self, config: Config) -> Optimizer:
```

Comment:
Theoretically, might the global and local optimizers be different for APFL? Maybe that's not admissible under their theory, but our previous implementation didn't enforce that they are the same.

One way we could handle this with a bit more generality is to rename `get_optimizer` to `get_optimizers` and have it return a dictionary of optimizers. For single optimizer implementations, this would just be keyed something like "optimizer" and for apfl (or other double optimizer approaches) this would be "global_optimizer" and "local_optimizer". That way we don't have to overload the idea of optimizer and they could be different. 

The property `self.optimizer` could also be renamed to `self.optimizers: Dict[str, torch.nn.Optimizer]` and basic client would simply used self.optimizers['optimizer'] to do the backprop, where as APFL would make use of two optimizers stored in the dictionary.

We could also get a little trickier and create classes with a base class returned for get_optimizer

Reply:
Updated in [61693e7](https://github.com/VectorInstitute/FL4Health/pull/62/commits/61693e7a371f5a3b4a0ab9237a87cc0dcf339b2d). Settled on a Union of optimizer and dict of str Optimizer as per our discussion on slack. Let me know if you have any issues with this.